### ISSUE TITLE
Does realm keep all objects of a transaction in native memory until commit?

### ISSUE THREAD

**Issue by ridcully99** [realm/realm-java#4644]
We are doing a synchronization of data from a server backend. In order to have a clean, consistent database all the time, we do this within a transaction and commit only after everything went well.

Now when getting too many items from the server (1.5 million items on first sync attempt), our app crashes with the following exception:

```
io.realm.internal.OutOfMemoryError: std::bad_alloc in /home/cc/repo/realm/release/realm/realm-library/src/main/cpp/io_realm_internal_Table.cpp line 669 
05-15 08:27:05.334 11067-11108/com.timetac.timetaclibrary.test W/System.err:     at io.realm.internal.Table.nativeSetLong(Native Method)
05-15 08:27:05.334 11067-11108/com.timetac.timetaclibrary.test W/System.err:     at io.realm.NodeToUserRealmProxy.insertOrUpdate(NodeToUserRealmProxy.java:1071)
05-15 08:27:05.334 11067-11108/com.timetac.timetaclibrary.test W/System.err:     at io.realm.LibraryRealmModuleMediator.insertOrUpdate(LibraryRealmModuleMediator.java:761)
05-15 08:27:05.334 11067-11108/com.timetac.timetaclibrary.test W/System.err:     at io.realm.Realm.insertOrUpdate(Realm.java:1170)
```
Does that mean that all items are kept in memory by realm, until the commit?
Can that be changed somehow? Are there any other ways or best practices for syncing?

**Comment by cmelchior** [comment]
Hi @ridcully99 Yes, everything is kept in memory until the final commit. Unfortunately, the only workaround is splitting the commit into multiple chunks.

**Comment by cmelchior** [comment]
Just out of curiosity. 1.5 million items is quite a lot. How many items you are sending over the network?

Would it instead be possible to create an assetfile you ship with the app instead?

**Comment by ridcully99** [comment]
Thanks for your quick answer. That is very sad though.

**Comment by cmelchior** [comment]
@ridcully99 I created #2614 to track this, but note that it will probably have a fairly low priority.

**Comment by Zhuinden** [comment]
you mean https://github.com/realm/realm-core/issues/2614

**Comment by cmelchior** [comment]
Ups, fixed 👍

**Comment by finnschiermer** [comment]
@ridcully99 Any idea what the memory consumption for those 1.5 mill objects are? If you put them in a realm file and compact the file, how large is it? Do you know how much physical memory is on the device where it fails?

**Comment by ridcully99** [comment]
The 1.5 million is just an example. It can also be many more items - this is a many-to-many connection table between users and tasks, so the number of items is more or less count(users) x count(tasks). Due to offline requirements we have to sync all that data to the device.

**Comment by ridcully99** [comment]
OK, 1 million of the items is about 28MB when compacted in the realm.

**Comment by finnschiermer** [comment]
Thx. As a workaround, I'd partition it into multiple commits as suggested earlier. I'd likely add a "meta" table for tracking the progress of these smaller transactions.

**Comment by cmelchior** [comment]
@ridcully99 28 MB doesn't sound like that much, to be honest. So I suspect something else is the problem here. How are you inserting the objects into Realm, using our streaming Json API or converting them to in-memory objects first?

Also, have you tried to attach a debugger during the process to see how the memory pressure in Java is?

The exception you are getting is definitely a C++ one, but it might just be an accident if you are holding on to way too many Java objects?

**Comment by Zhuinden** [comment]
Building many-many relations is kinda tricky because memory requirement scales as O(N).

Single objects can easily be inserted with O(1) memory requirement (on Java side, anyways).

**Comment by ridcully99** [comment]
Here is a test that produces the crash. The `createObservable() `method simulates the backend server and produces a new 'page' of data every 500ms.

```
system_process I/ActivityManager:   sys   P    52534 kB: system (pid 1573) fixed
                                                                 pers  P    38342 kB: com.android.systemui (pid 1694) fixed
                                                                 pers  P    16683 kB: com.android.phone (pid 1914) fixed
                                                                 fore  SF  677543 kB: com.timetac.timetaclibrary.test (pid 2699) instrumentation
                                                                 fore  T    25724 kB: com.android.launcher3 (pid 2361) top-activity
                                                                 prcp  IB    9979 kB: com.android.inputmethod.latin (pid 1898) service
                                                                                      com.android.inputmethod.latin/.LatinIME<=Proc{1573:system/1000}
                                                                 prev  LA   29529 kB: com.android.settings (pid 2156) previous
                                                                           939445 kB: TOTAL
```


```
@Test
    public void testSpeedInLoop_singleTransaction() {

        final Realm[] tempRealm = new Realm[1];

        Completable.fromAction(() -> {
            tempRealm[0] = Realm.getInstance(getRealmConfiguration());
            tempRealm[0].executeTransaction(r -> r.deleteAll());
            tempRealm[0].beginTransaction();
            }).subscribeOn(Schedulers.single()).blockingAwait();

        boolean[] complete = new boolean[]{false};

        createObservable().subscribeOn(Schedulers.io())
            //.doOnEvent((ignored, ignored2) -> parallelCount[0]--)
            .observeOn(Schedulers.single())
            .doOnTerminate(() -> complete[0] = true)
            .subscribe(
                response -> {
                    long a = System.currentTimeMillis();
                    tempRealm[0].insertOrUpdate(response);
                    long b = System.currentTimeMillis();
                    Log.d(TAG, "insert took " + (b - a) + "ms");
                }
            );

        while (!complete[0]) {
            try {
                Thread.sleep(50);
            } catch (InterruptedException ignored) {
            }
        }

        Completable.fromAction(() -> {
            long a = System.currentTimeMillis();
            tempRealm[0].commitTransaction();
            long b = System.currentTimeMillis();
            Log.d(TAG, "commit took " + (b - a) + "ms");
            tempRealm[0].close();
            Realm.compactRealm(getRealmConfiguration());
            Log.d(TAG, "Realm file name " + getRealmConfiguration().getPath());
        }).subscribeOn(Schedulers.single()).blockingAwait();
    }

    private Observable<List<NodeToUser>> createObservable() {
        return Observable.create(e -> {
            int start = 0;
            while (start < TEST_AMOUNT) {
                Log.d(TAG, "start = " + start);
                e.onNext(createNodeToUsers(start, 10000));
                Thread.sleep(500);
                start += 10000;
            }
            e.onComplete();
        });
    }

    private List<NodeToUser> createNodeToUsers(int start, int n) {
        List<NodeToUser> result = new ArrayList<>(n);
        while (n > 0) {
            NodeToUser item = new NodeToUser();
            item.setUser(1);
            item.setNode(start + n);
            item.buildPrimaryKey();
            result.add(item);
            n--;
        }
        return result;
    }

    public RealmConfiguration getRealmConfiguration() {
        return
            new RealmConfiguration.Builder()
                .name("testrealm")
                .modules(new LibraryRealmModule())
                .build();
    }

    private class NodeToUser extends RealmObject {
        @PrimaryKey
        private String nodeAndUser;

        @SerializedName("node")
        @Expose
        //@Index
        private Integer node;
        @SerializedName("user")
        @Expose
        //@Index
        private Integer user;
        @SerializedName("is_favourite")
        @Expose
        private int isFavourite;
        @SerializedName("is_todo")
        @Expose
        private int isTodo;
        @SerializedName("access")
        @Expose
        private int access;

        // -- Additional fields needed...
        private Project project;
        private Task task;
        private Integer sortOrder;

        public void buildPrimaryKey() {
            nodeAndUser = node + "§" + user;
        }

        public Integer getNode() {
            return node;
        }

        public void setNode(Integer node) {
            this.node = node;
        }

        public Integer getUser() {
            return user;
        }

        public void setUser(Integer user) {
            this.user = user;
        }

        public int getIsFavourite() {
            return isFavourite;
        }

        public void setIsFavourite(int isFavourite) {
            this.isFavourite = isFavourite;
        }

        public int getIsTodo() {
            return isTodo;
        }

        public void setIsTodo(int isTodo) {
            this.isTodo = isTodo;
        }

        public Integer getAccess() {
            return access;
        }

        public void setAccess(Integer access) {
            this.access = access;
        }

        public Project getProject() {
            return project;
        }

        public void setProject(Project project) {
            this.project = project;
        }

        public Task getTask() {
            return task;
        }

        public void setTask(Task task) {
            this.task = task;
        }

        public Integer getSortOrder() {
            return sortOrder;
        }

        public void setSortOrder(Integer sortOrder) {
            this.sortOrder = sortOrder;
        }
    }
```

**Comment by cmelchior** [comment]
```
    @Test
    public void largeTransaction() {
        int testObjects = 10000000; // 10.000.000
        int batch = 10000;
        realm.beginTransaction();

        int start = 0;
        while (start < testObjects) {
            // Create test objects in batches of 10.000 at a time
            // until the limit is hit
            realm.insertOrUpdate(createNodeToUsers(start, batch));
            start += batch;
        }
        realm.commitTransaction();
        assertEquals(testObjects, realm.where(NodeToUser.class).count());
    }
```

This more simplified unit test crashes with the same exception, but Android Studio is refusing to attach the memory analyzer for some reason so I cannot verify if it behaves the same.

**Comment by bmeike** [comment]
Related issue: https://github.com/realm/realm-java/issues/4657

**Comment by bmeike** [comment]
This one seems to be much more reasonably behaved that others we've seen...
![profile](https://cloud.githubusercontent.com/assets/277848/26183550/52b9e6fa-3b34-11e7-97cd-f926c6863945.png)

**Comment by bmeike** [comment]
Also related: https://github.com/realm/realm-core/issues/1467

**Comment by bmeike** [comment]
This looks absolutely plausible, to me.  The heap is dominated by the objects being inserted into the DB and the graph shows that most of the memory consumption is native (that is, probably, in the DB).

<img width="1893" alt="crash" src="https://cloud.githubusercontent.com/assets/277848/26420004/52172536-4076-11e7-808e-49027178a998.png">

Closing as a duplicate of https://github.com/realm/realm-core/issues/1467
